Library/OcAppleKernelLib/KextPatcher.c

```diff
@@ -282,23 +365,28 @@ PatcherBlockKext (
     return EFI_UNSUPPORTED;
   }
 
-  KmodOffset = Context->VirtualKmod - Context->VirtualBase;
-  KmodInfo   = (KMOD_INFO_64_V1 *)((UINT8 *) MachoGetMachHeader64 (&Context->MachContext) + KmodOffset);
-  StartAddr  = KcFixupValue (KmodInfo->StartAddr, NULL);;
+  KmodOffset = Context->VirtualKmod - Context->VirtualBase + Context->FileOffset;
+  KmodInfo   = (UINT8 *) MachoGetMachHeader (&Context->MachContext) + KmodOffset;
+  if (Context->Is32Bit) {
+    StartAddr = ((KMOD_INFO_32_V1 *) KmodInfo)->StartAddr;
+  } else {
+    StartAddr = ((KMOD_INFO_64_V1 *) KmodInfo)->StartAddr;
+    StartAddr = KcFixupValue (StartAddr, NULL);
+  }
 
-  if (OcOverflowAddU64 (KmodOffset, sizeof (KMOD_INFO_64_V1), &TmpOffset)
+  if (OcOverflowAddU64 (KmodOffset, Context->Is32Bit ? sizeof (KMOD_INFO_32_V1) : sizeof (KMOD_INFO_64_V1), &TmpOffset)
     || TmpOffset > MachoGetFileSize (&Context->MachContext)
     || StartAddr == 0
     || Context->VirtualBase > StartAddr) {
     return EFI_INVALID_PARAMETER;
   }
 
-  TmpOffset = StartAddr - Context->VirtualBase;
+  TmpOffset = StartAddr - Context->VirtualBase + Context->FileOffset;
```

Comment:
This needs overflow checks now as adding `Context->FileOffset` may overflow in theory.